src/libmongoc/src/mongoc/mongoc-http.c

```diff
@@ -115,7 +119,11 @@ _mongoc_http_send (const mongoc_http_request_t *req,
       goto fail;
    }
 
-   stream = mongoc_client_connect_tcp (timeout_ms, &host_list, error);
+   stream = mongoc_client_connect_tcp (
+      // +1 to prevent passing zero as a timeout
+      mcd_get_milliseconds (mcd_timer_remaining (timer)) + 1,
+      &host_list,
+      error);
```

Comment:
Good catch. If `_mongoc_http_send` requires a non-zero timeout, I suggest checking the timeout before each blocking call:

```c
#define CHECK_TIMEOUT(msg)                                        \
   if (mcd_get_milliseconds (mcd_timer_remaining (timer)) == 0) { \
      bson_set_error (error,                                      \
                      MONGOC_ERROR_STREAM,                        \
                      MONGOC_ERROR_STREAM_SOCKET,                 \
                      "Timeout before %s: %s",                    \
                      msg,                                        \
                      req->host);                                 \
      goto fail;                                                  \
   }

   CHECK_TIMEOUT ("connect")
```

`mongoc_stream_tls_handshake_block` interprets 0 as "no timeout".

Reply:
Looking into this, there's a several goofs in timing that may be a problem, e.g. an unguarded multiply of `timeout_ms * 1000` that can overflow. Requiring the timeout to not be zero is strange behavior.

I also realize now that asserting on overflow in the duration code is probably bad (for example, one might want `mcd_seconds(INT64_MAX)` to become a maximum duration, rather than terminate the program, and using assertions implies that the caller is supposed to know what the time encoding is).